torchvision/transforms/_functional_tensor.py

```diff
@@ -302,21 +302,31 @@ def _hsv2rgb(img: Tensor) -> Tensor:
     h, s, v = img.unbind(dim=-3)
     i = torch.floor(h * 6.0)
     f = (h * 6.0) - i
-    i = i.to(dtype=torch.int32)
+    i = i.to(dtype=torch.long)
```

Comment:
@nlgranger what is the reason to pass to int64 here instead of keeping int32 ?

Reply:
It is now used for tensor indexing.

Reply:
I think we can also index with int32, isn't the case somewhere ?
I'm checking your code and running with int32 I do not see any errors on cpu or cuda.
Also, maybe it could be also possible to call gather single time instead of 3 times...

Reply:
It used to be that only int64 indices were supported but I guess this is not the case anymore. However if they get cast implicitly anyway I might as well leave the long conversion.

To fuse the indices, I think you need something like that:
```py
imh, imw = img.shape[-2], img.shape[-1]
vpqt = torch.stack((v, p, q, t), dim=-3).view(-1, 4, imh * imw)
imb = vpqt.shape[0]

select = select.view(1, 3, 6).expand(imb, 3, 6).gather(2, i.view(imb, 1, imh * imw).expand(imb, 3, imh * imw))
return vpqt.gather(1, select).view(img.shape)
```

It is a lot less readable IMHO, not faster and has a higher peak memory usage due to storing the three indices simultaneously.

Reply:
There is another more simple way to do that:
```python
    vpqt = torch.stack((v, p, q, t), dim=-3)

    # vpqt -> rgb mapping based on i
    select = torch.tensor(
        [[0, 2, 1, 1, 3, 0], [3, 0, 0, 2, 1, 1], [1, 1, 3, 0, 0, 2]],
        dtype=torch.long, device=img.device
    )

    select = select[:, i]
    if select.ndim > 3:
        select = select.transpose(0, 1)
    return vpqt.gather(-3, select)
```

Reply:
Indeed, I was overthinking it to keep the indices contiguous. I have forced-pushed your suggestion but kept the pin memory then `.to()` steps. It helps [avoid CPU/GPU syncs](https://github.com/pytorch/pytorch/issues/98673).